### ISSUE TITLE
GitLab.com as a source code service

### ISSUE THREAD

**Issue by alexsorokoletov** [microsoft/appcenter#37]
**What App Center service does this feature apply to?**
We keep a lot of our projects in gitlab.com (online edition).
We use AppCenter on any project that is hosted in VSTS or Bitbucket, however most of our projects are still in GitLab.

**Describe the solution you'd like**
We want to be able to build an iOS/Android/Xamarin apps from source code hosted on gitlab.com

**Describe alternatives you've considered**
Alternatives we considered: migrating to VSTS or Bitbucket (these services unfortunately are not on par with GitLab and we will loose a lot of features we use daily for our workflow); using another build service (we use a hosted Jenkins for now - works okay but it is clear future is with AppCenter-like services).

**Additional context**
I will be glad to provide any additional information

**Comment by nrajpurkar** [comment]
@alexsorokoletov thanks for creating an issue for this here. I understand the additional work required in moving your code elsewhere. We will keep this issue open to track further interest in providing support for using GitLab.com as a source provider.

**Comment by alexsorokoletov** [comment]
Thank you!
How much interest do you need?
How many retweets ;) to make it happen?

Alex

**Comment by patniko** [comment]
Depends on the size of the request. This would require a good number of changes and planning so we would need a good amount of reactions on the issue for us to drop other plans and prioritize. Fair to say that 100 thumbs up would force us to take notice for an effort this big.

**Comment by alexsorokoletov** [comment]
Fair enough, thank you!

**Comment by Kondamon** [comment]
+1

**Comment by markbeaton** [comment]
+1 We’d love GitLab support too, although we self-host our own Omnibus edition under our own domain.

**Comment by nrajpurkar** [comment]
@markbeaton thanks for the clarification. Supporting GitLab.com in the cloud and supporting self-hosted on-premise GitLab looks a little different for us so it is great to understand what the request is actually for.

**Comment by alexshikov** [comment]
All these years we are waiting for the GitLab support with no luck. Bitrise makes a great job for us all this time. But it would be awesome to start new projects on AppCenter ;)

To be honest, I doubt, 100 upvotes make this feature move forward, but will see soon :)

**Comment by patniko** [comment]
At 100 upvotes I promise you we'll take a serious look at what it would really take to get done for gitlab.com and circle back with a real update from our team @alexshikov. Nothing is off limits if enough people want it, but we do keep supportability and our roadmap in mind when deciding on new integrations.

**Comment by Bilonik** [comment]
+1

**Comment by cnnranderson** [comment]
Looking for Enterprise support, ourselves! Crossing our fingers for support 🤞 

+1

**Comment by kijanawoodard** [comment]
What's the work around for gitlab users?
Currently, I push to hockeyapp via curl.

**Comment by Kondamon** [comment]
@kijanawoodard some people mirror their repository to Github as a workaround to be able to use Appcenter.

**Comment by kijanawoodard** [comment]
@Kondamon ouch. That won't work in this case. We use private, on prem gitlab. We push via the hockey api. Is there an equivalent api endpoint for appcenter?

I was just browsing the appcenter swagger and nothing popped out as obvious, but I'm on a small screen atm.

**Comment by nrajpurkar** [comment]
@kijanawoodard I think I'm hearing two different requests here. This original issue is about using GitLab.com as a source code provider to be able to build your apps using App Center's Build service. For that, the workaround is like @Kondamon said of mirroring a repo. It sounds like your question is you distribute your app though - is that correct? In that case, you could use the [Distribute APIs](https://openapi.appcenter.ms/#/distribute) for App Center regardless of where you build your app.

**Comment by nrajpurkar** [comment]
@kijanawoodard In addition to that, supporting GitLab.com online and supporting on-prem source code providers looks a little bit different to us and at this time we don't support any on-prem source code providers.

**Comment by patniko** [comment]
@kijanawoodard I would recommend using the `appcenter-cli` to ease uploads for now. Our API's support much larger files than HockeyApp with the addition of chunking, but this comes with complexity. We're tracking a work item right now to make the API calls for uploading simpler in App Center.

**Comment by drobs** [comment]
+1 Gitlab support for on premise would be awesome

**Comment by ateek** [comment]
+1 Gitlab support

**Comment by KSemenenko** [comment]
Add private repositories too. We use gitlab,  and there no way to use builds in appcenter.

**Comment by markbeaton** [comment]
@nrajpurkar I was under the (possibly mistaken) impression that GitLab.com and self-hosted GitLab installations were running the same software, so shouldn't work any differently other than the base URLs being different (i.e. https://gitlab.com vs https://gitlab.mycompany.com).

**Comment by AdamZaczek** [comment]
Make it happen

**Comment by filiphosko** [comment]
+1 Would like to see GitLab support too, at this moment we have to mirror our React Native projects to BitBucket.

**Comment by patniko** [comment]
Looks like we hit 100 reactions. As promised, @nrajpurkar and I will start taking a hard look in May's iteration to see what we can for gitlab.com support and circle back here.

**Comment by waelbouchnak** [comment]
+1 Gitlab support

**Comment by thetanz-geoff** [comment]
+1 GitLab support, please!

**Comment by bendada3** [comment]
+1

**Comment by claudiocleberson** [comment]
+1

**Comment by malhobayyeb** [comment]
+1 because all of our projects are hosted in GitLab.

**Comment by ka-lex** [comment]
+1 we need Gitlab support :)

**Comment by prabakaranr** [comment]
+1

**Comment by karthikeyanvis** [comment]
+1

**Comment by nrajpurkar** [comment]
As we're taking a closer look into supporting GitLab.com, we'd love to better understand the ask here. Please react to this comment with the Hooray react for general GitLab.com support and the Rocket react for self-hosted GitLab.mycompany.com support. Looking forward to working toward solving this feature request!

**Comment by fballan93** [comment]
> As we're taking a closer look into supporting GitLab.com, we'd love to better understand the ask here. Please react to this comment with the _Hooray react_ for **general GitLab.com support** and the _Rocket react_ for **self-hosted GitLab.mycompany.com support**. Looking forward to working toward solving this feature request!

I was super frustrated looking for a way to get my app build automated but it was always a hassle to keep moving from gitlab to GitHub . THANKKKK YOUUU <3

**Comment by caseyjmorton** [comment]
+1 for git support

**Comment by nrajpurkar** [comment]
We mentioned in our May iteration plan (#361) that we'd be investigating what support for GitLab.com would look like. After taking a further look into this, we will be beginning work on implementing general GitLab.com support as a part of our June iteration plan (#466). This may not be completed in June as this isn't a small change on our end, and we'll be working on this in parallel with other App Center priorities. We'll keep you updated as we make progress!

**Comment by prabakaranr** [comment]
Is there an update to support the self-hosted gitlab repositories? Do we need to create a separate issue for this?

**Comment by DennisPan** [comment]
@prabakaranr We plan to run a small experiment with self-hosted gitlab repositories, once we finish the work of supporting GitLab.com, given that we see a fair percentage of the interest comes from self-hosted. If we end up deciding to not move forward, then we can create a separate issue to track the interest. That would be my suggestion.

**Comment by DennisPan** [comment]
For everyone who are interested in **GitLab.com** support, if you would like to participate in an early preview, please reach out to me via email (can be found in my profile). Currently looking at finishing the initial integration in about a week. We'd love to have you start trying and collect feedbacks!

**Comment by caseyjmorton** [comment]
+1 for GitLab on prem.

**Comment by lyricyu** [comment]
For everyone who is interested in **self-hosted GitLab.mycompany.com** support, we would like to get a better understanding of your interest in this support. If you would like to contribute to this, please reach out to me via email (my email can be found in my profile). We'd love to hear from you to get a better idea of how to proceed with this feature!

**Comment by lyricyu** [comment]
We are very excited to announce that we have just **officially enabled support for GitLab.com**! 🎉 We're looking forward to seeing your projects hosted in GitLab come to life, and can't wait to hear your feedback!

**Comment by filiphosko** [comment]
It's awesome, thanks!

**Comment by WebDucer** [comment]
Now support for on perm and as issue tracker, than it is perfect.

**Comment by artem1989** [comment]
How to set up company gitlab account with AppCenter? It only follows with my private...

**Comment by DennisPan** [comment]
@artem1989 Could you help me understand what do you mean by "company gitlab account"? Are you referring to self-hosted GitLab? If so, it is expected that this isn't supported yet, and we are working on that support

**Comment by artem1989** [comment]
Yes, I meant on-prem gitlab e.g. "main.gitlab.mycompany.com"

**Comment by artem1989** [comment]
When is it approximately expected?

**Comment by DennisPan** [comment]
@artem1989 Most likely couple weeks out.

**Comment by mseyad** [comment]
For everyone who is interested in GitLab on-prem support, if you would like to participate in an early preview, please reach out to me via email (can be found in my profile). Currently looking at finishing the initial integration in about a week. We'd love to have you start trying and collect feedback!

**Comment by JourKung** [comment]
Great thank you!

**Comment by nrajpurkar** [comment]
We've rolled out self-hosted GitLab support for all! Take a look at some [documentation](https://docs.microsoft.com/en-us/appcenter/build/connect#connecting-to-self-hosted-gitlab-instances) and test it out! Closing out this issue after a few months of work 👊

**Comment by LeeJCherry** [comment]
Any support for IP Whitelisting?  We have self-hosted Gitlab only accessible via VPN or IP Whitelist. Where can I find the ranges to whitelist?

**Comment by sam4995** [comment]
Hi @nilofer / @DennisPan  ,

We are facing an issue connecting the self-hosted company GitLab instance.
We have configured the SSL certificate for the GitLab instance that matches exactly. Example: https://gitlab.my-company.com/.

Still while connecting the from AppCenter we are getting the error 'Something went wrong, please try again.'
We have whitelisted all required URLs including 'build.appcenter.ms'.

![image](https://user-images.githubusercontent.com/70276312/143240304-4daa0d96-d8dd-4f22-91db-85cdbe69322b.png)

A help will be much appreciated, as we are stuck on this for weeks.

Regards,
Sampras Dias
